### PR TITLE
Fast tmx (added culling for Tile Map)

### PR DESCRIPTION
**Pull request by dabingnn**
Please do not merge it now. 
Feedbacks and comments are appreciated.
Performance tests and improvements is needed.

### CODE REVIEW

File: cocos/2d/CCFastTMXLayer.cpp [hunk 1]
```diff
@@ -0,0 +1,777 @@
...
+
+    //clear quadsNumber
+    _quadsNumber.clear();
+
+    Size texSize = _tileSet->_imageSize;
+    for (int y =  visibleTiles.origin.y - tilesOverY; y < visibleTiles.origin.y + visibleTiles.size.height + tilesOverY; ++y)
+    {
+        if(y<0 || y >= _layerSize.height)
+            continue;
+        for (int x = visibleTiles.origin.x - tilesOverX; x < visibleTiles.origin.x + visibleTiles.size.width + tilesOverX; ++x)
```
Comment (gelldur) [inline cocos/2d/CCFastTMXLayer.cpp]: Why this for is so big :/ could you refactor it?

File: cocos/2d/CCFastTMXLayer.cpp [hunk 2]
```diff
@@ -0,0 +1,777 @@
...
+                quad->tr.colors = Color4B::WHITE;
+                
+                tilesUsed++;
+            }
+            
+            if (tilesUsed >= MAX_QUADS_COUNT)
+            {
+                break;
+            }
+        } // for x
```
Comment (gelldur) [inline cocos/2d/CCFastTMXLayer.cpp:277]: If you refactor for you don;t have to do such things

File: cocos/2d/CCFastTMXLayer.cpp [hunk 3]
```diff
@@ -0,0 +1,777 @@
...
+    _texture->setAliasTexParameters();
+
+    //CFByteOrder o = CFByteOrderGetCurrent();
+
+    // Parse cocos2d properties
+    this->parseInternalProperties();
+
+    Size screenSize = Director::getInstance()->getWinSize();
+
+    switch (_layerOrientation)
```
Comment (gelldur) [inline cocos/2d/CCFastTMXLayer.cpp:333]: default missing. Assert or something?

File: cocos/2d/CCFastTMXLayer.cpp [hunk 4]
```diff
@@ -0,0 +1,777 @@
...
+    if (_properties.find(propertyName) != _properties.end())
+        return _properties.at(propertyName);
+    
+    return Value();
+}
+
+void FastTMXLayer::parseInternalProperties()
+{
+    auto vertexz = getProperty("cc_vertexz");
+    if (!vertexz.isNull())
```
Comment (gelldur) [inline cocos/2d/CCFastTMXLayer.cpp]: why you don't do such thing

if(vertexz.isNull() == nullptr)
{
    return;
}

you will save indention

File: cocos/2d/CCFastTMXLayer.cpp [hunk 5]
```diff
@@ -0,0 +1,777 @@
...
+void FastTMXLayer::setTileGID(int gid, const Vec2& tileCoordinate, TMXTileFlags flags)
+{
+    CCASSERT(tileCoordinate.x < _layerSize.width && tileCoordinate.y < _layerSize.height && tileCoordinate.x >=0 && tileCoordinate.y >=0, "TMXLayer: invalid position");
+    CCASSERT(_tiles, "TMXLayer: the tiles map has been released");
+    CCASSERT(gid == 0 || gid >= _tileSet->_firstGid, "TMXLayer: invalid gid" );
+    
+    TMXTileFlags currentFlags;
+    int currentGID = getTileGIDAt(tileCoordinate, &currentFlags);
+    
+    if (currentGID != gid || currentFlags != flags)
```
Comment (gelldur) [inline cocos/2d/CCFastTMXLayer.cpp]: again you can save indention

File: cocos/2d/CCFastTMXTiledMap.cpp [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+    layer->setupTiles();
+
+    return layer;
+}
+
+TMXTilesetInfo * FastTMXTiledMap::tilesetForLayer(TMXLayerInfo *layerInfo, TMXMapInfo *mapInfo)
+{
+    Size size = layerInfo->_layerSize;
+    auto& tilesets = mapInfo->getTilesets();
+    if (tilesets.size()>0)
```
Comment (gelldur) [inline cocos/2d/CCFastTMXTiledMap.cpp]: you can save a lot of indentions

File: cocos/2d/CCFastTMXLayer.h [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+     * @lua NA
+     */
+    virtual ~FastTMXLayer();
+
+    /** returns the tile gid at a given tile coordinate. It also returns the tile flags.
+     */
+    int getTileGIDAt(const Vec2& tileCoordinate, TMXTileFlags* flags = nullptr);
+    CC_DEPRECATED_ATTRIBUTE int tileGIDAt(const Vec2& tileCoordinate, TMXTileFlags* flags = nullptr){
+        return getTileGIDAt(tileCoordinate, flags);
+    };
```
Comment (minggo) [inline cocos/2d/CCFastTMXLayer.h]: Why new class need deprecated functions?

File: cocos/2d/CCFastTMXTiledMap.cpp [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+    {
+        TMXTilesetInfo* tileset = nullptr;
+        for (auto iter = tilesets.crbegin(); iter != tilesets.crend(); ++iter)
```
Comment (minggo) [inline cocos/2d/CCFastTMXTiledMap.cpp]: Could use for loop.

File: cocos/2d/CCFastTMXTiledMap.h [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+class TMXTilesetInfo;
+class TMXMapInfo;
+
+/**
+ * @addtogroup tilemap_parallax_nodes
+ * @{
+ */
+
+/** Possible orientations of the TMX map */
+enum
```
Comment (minggo) [inline cocos/2d/CCFastTMXTiledMap.h]: Please use `enum class`.

File: cocos/2d/CCFastTMXTiledMap.h [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+{
+    /** Orthogonal orientation */
+    FastTMXOrientationOrtho,
+
+    /** Hexagonal orientation */
+    FastTMXOrientationHex,
+
+    /** Isometric orientation */
+    FastTMXOrientationIso,
```
Comment (minggo) [inline cocos/2d/CCFastTMXTiledMap.h]: Elements should use capital words, and use `_` to connect words.

File: cocos/2d/CCFastTMXTiledMap.h [hunk 11]
```diff
@@ -0,0 +1,221 @@
...
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+****************************************************************************/
+#ifndef __CCTMX_TILE_MAP2_H__
+#define __CCTMX_TILE_MAP2_H__
```
Comment (minggo) [inline cocos/2d/CCFastTMXTiledMap.h]: Don't match file name.

File: cocos/2d/CCFastTMXTiledMap.h [hunk 12]
```diff
@@ -0,0 +1,221 @@
...
+
+    /** object groups */
+    inline const Vector<TMXObjectGroup*>& getObjectGroups() const { return _objectGroups; };
+    inline Vector<TMXObjectGroup*>& getObjectGroups() { return _objectGroups; };
+    inline void setObjectGroups(const Vector<TMXObjectGroup*>& groups) {
+        _objectGroups = groups;
+    };
+    
+    /** properties */
+    inline ValueMap& getProperties() { return _properties; };
```
Comment (minggo) [inline cocos/2d/CCFastTMXTiledMap.h]: const?

File: cocos/2d/CCFastTMXTiledMap.h [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+
+Each property is stored as a key-value pair in an MutableDictionary.
+You can obtain the properties at runtime by:
+
+map->getProperty(name_of_the_property);
+layer->getProperty(name_of_the_property);
+objectGroup->getProperty(name_of_the_property);
+object->getProperty(name_of_the_property);
+
+@since v0.8.1
```
Comment (minggo) [inline cocos/2d/CCFastTMXTiledMap.h]: since 3.2

**Comment by minggo** [comment]
Please fix conflict.
Thanks.

File: cocos/2d/CCFastTMXLayer.cpp (hunk 3 above)
Comment (minggo) [inline cocos/2d/CCFastTMXLayer.cpp:333]: It is fixed.

**Pull request merged**